revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,293 @@
+// length of this array will always be equal to the length of the certificate
+// chain.
+func CheckStatus(opts Options) []*base.CertRevocationResult {
+	result := make([]*base.CertRevocationResult, len(opts.CertChain))
+
+	if len(opts.CertChain) == 0 {
+		return result
+	}
+
+	// Validate cert chain structure
+	if len(opts.CertChain) == 1 {
+		if err := validateIssuer(opts.CertChain[0], opts.CertChain[0]); err != nil {
+			fillResultsWithError(result, base.InvalidChainError{IsInvalidRoot: true, Err: err})
+			return result
+		}
+	} else {
+		if err := coreX509.ValidateCodeSigningCertChain(opts.CertChain, &opts.SigningTime); err != nil {
+			fillResultsWithError(result, base.InvalidChainError{Err: err})
+			return result
+		}
+	}
+
+	// Check status for each cert in cert chain
+	var wg sync.WaitGroup
+	for i, cert := range opts.CertChain[:len(opts.CertChain)-1] {
+		wg.Add(1)
+		// Assume cert chain is accurate and next cert in chain is the issuer
+		go func(i int, cert *x509.Certificate) {
+			defer wg.Done()
+			result[i] = certCheckStatus(cert, opts.CertChain[i+1], opts)
+		}(i, cert)
+	}
+	// Last is root cert, which will never be revoked by OCSP
+	result[len(opts.CertChain)-1] = &base.CertRevocationResult{
+		Result: base.NonRevokable,
+		ServerResults: []*base.ServerResult{{
+			Result: base.NonRevokable,
+			Error:  nil,
+		}},
+	}
+
+	wg.Wait()
+	return result
+}
+
+func fillResultsWithError(results []*base.CertRevocationResult, err error) {
+	for i := range results {
+		results[i] = &base.CertRevocationResult{
+			Result: base.Unknown,
+			ServerResults: []*base.ServerResult{{
+				Result: base.Unknown,
+				Error:  err,
+			}},
+		}
+	}
+}
+
+func validateIssuer(subject *x509.Certificate, issuer *x509.Certificate) error {
+	if err := subject.CheckSignature(subject.SignatureAlgorithm, subject.RawTBSCertificate, subject.Signature); err != nil {
+		return err
+	}
+	if !bytes.Equal(issuer.RawSubject, subject.RawIssuer) {
+		return errors.New("parent's subject does not match issuer for a cert in the chain")
+	}
+	return nil
+}
+
+func certCheckStatus(cert, issuer *x509.Certificate, opts Options) *base.CertRevocationResult {
+	ocspURLs := cert.OCSPServer
+	if len(ocspURLs) == 0 {
+		// OCSP not enabled for this certificate.
+		return &base.CertRevocationResult{
+			Result:        base.NonRevokable,
+			ServerResults: []*base.ServerResult{errToServerResult(NoOCSPServerError{})},
+		}
+	}
+
+	serverResults := make([]*base.ServerResult, len(ocspURLs))
+	for serverIndex, server := range ocspURLs {
+		serverResult := checkStatusFromServer(cert, issuer, server, opts)
+		switch t := serverResult.Error.(type) {
+		case nil, RevokedError, UnknownStatusError:
+			// A valid response has been received from an OCSP server
+			// Result should be based on only this response, not any errors from
+			// other servers
+			return serverResultsToCertRevocationResult([]*base.ServerResult{errToServerResult(t)})
```

Comment:
> I think since there is no guidance on an approach here, it may be best to just call out this behavior in the spec (noting that the first valid response is used for efficiency). In the future, we may consider an option to allow users to force the check for all valid servers and use the valid result with a thisUpdate time closest to the current time.

Yeah, that's my concern, because without a clear guidance, each user might have their own version of understanding on it. IMO, we should record the current behavior in notation's spec for users' reference.

Reply:
Same here, we might need to open an issue for this in https://github.com/notaryproject/notaryproject.